File: Lista de exercicio/projetoJAVA/src/projetojava/questao4.java
package projetojava;

abstract class Animal {
    abstract String som();
    abstract String comportamento();
}

class Cachorro extends Animal {
    @Override
    String som() {
        return "Au Au";
    }

    @Override
    String comportamento() {
        return "O cachorro corre atrás do próprio rabo.";
    }
}

class Gato extends Animal {
    @Override
    String som() {
        return "Miau";
    }

    @Override
    String comportamento() {
        return "O gato destrói toda a casa fingindo que nada aconteceu.(maligno)";
    }
}

class Pato extends Animal {
    @Override
    String som() {
        return "Quack";
    }

    @Override
    String comportamento() {
        return "O pato nada no lago.";
    }
}

class Pinto extends Animal {
    @Override
    String som() {
        return "Piu";
    }

    @Override
    String comportamento() {
        return "O pinto sai do ovo.";
    }
}

public class questao4 {
    public static void main(String[] args) {
        Animal[] animais = {new Cachorro(), new Gato(), new Pato(), new Pinto()};

        for (Animal animal : animais) {
            System.out.println(animal.getClass().getSimpleName() + ":");
            System.out.println("Som: " + animal.som());
            System.out.println("Comportamento: " + animal.comportamento() + "\n");
        }
    }
}

File: Lista de exercicio/projetoPYTHON/questao4/questao4.py
class Animal:
    def som(self):
        raise NotImplementedError("Método deve ser implementado na subclasse")

    def comportamento(self):
        raise NotImplementedError("Método deve ser implementado na subclasse")

class Cachorro(Animal):
    def som(self):
        return "Au Au"

    def comportamento(self):
        return "O cachorro corre atras do proprio rabo."

class Gato(Animal):
    def som(self):
        return "Miau"

    def comportamento(self):
        return "O gato destroi toda a casa fingindo que nada aconteceu (bicho maligno)."

class Pato(Animal):
    def som(self):
        return "Quack"

    def comportamento(self):
        return "O pato nada no lago."

class pinto(Animal):
    def som(self):
        return "piu"

    def comportamento(self):    
        return "o pinto sai do ovo"

def executar_questao4():
    animais = [Cachorro(), Gato(), Pato(), pinto()]

    for animal in animais:
        print(f"{animal.__class__.__name__}:")
        print(f"Som: {animal.som()}")
        print(f"Comportamento: {animal.comportamento()}\n")
